spatial-tracking-explainer.md

```diff
@@ -0,0 +1,382 @@
+  }
+  // Close the loop
+  let pointA = xrFrameOfReference.boundsGeometry[pointCount-1];
+  let pointB = xrFrameOfReference.boundsGeometry[0];
+  boundsMesh.addQuad(
+        pointA.x, 0, pointA.z, // Quad Corner 1
+        pointB.x, 2.0, pointB.z) // Quad Corner 2
+  }
+}
+```
+
+### Unbounded Frame of Reference
+A _unbounded_ experience is one in which the user is able to freely move around their physical environment. An example of this sort of experience would be a campus tour or a whole-house renovation preview. These experiences explicitly require that the user be unbounded in their ability to walk around, and the unbounded frame of reference will adjust its origin as needed to maintain optimal stability for the user, even if the user walks many meters from the origin. In doing so, the origin may drift from its original physical location.
+
+> **Note regarding XRCoordinateSystem**
+>
+> When building an _unbounded_ experience, developers may additionally desire to "lock" content to a specific physical location to prevent it from drifting as users travel beyond a few meters; this functionality is known as _anchors_.  In addition to _anchors_, today's XR platforms offer other environment-related features such as _spatial meshes_, _point clouds_, _markers_, and more. While none of these features are yet part of the WebXR Device API, there are proposed designs under active development.
+>
+> The common property of these additional features and `XRFrameOfReference`, is that they all act as _spatial roots_ that are independently tracked by the underlying tracking systems. The concept of a _spatial roots_ is represented in the WebXR Device API as an `XRCoordinateSystem`.  Each instance of an `XRCoordinateSystem` does not have a fixed relationship with any other.  Instead, on a frame-by-frame basis, the tracking systems must attempt to locate them and compute their relative locations.  
+
+```js
+let xrSession = null;
+let xrFrameOfReference = null;
+
+function onSessionStarted(session) {
+  xrSession = session;
+  xrSession.requestFrameOfReference({ type:'unbounded' })
+  .then((frameOfReference) => {
+    xrFrameOfReference = frameOfReference;
+  })
+  .then(setupWebGLLayer)
+  .then(() => {
+    xrSession.requestAnimationFrame(onDrawFrame);
+  });
+}
+```
+
+As developers start building experiences with the potential to span kilometers, they will also need to deal with the fact that floating point precision issues will be introduced.  When this is close to occurring, developers should create a new `XRUnboundedFrameOfReference`,and use the `coordinateSystem.getTransformTo()` method to reparent any nearby virtual object to the new frame of reference.  However, practically speaking, large _unbounded_ experiences will often segment app/game assets and logic into discrete map segments and, depending on map segment size, this may also be an opportune time to create a new `XRUnboundedFrameOfReference`.
```

Comment:
We talked about this at the F2F a bit, and I've discussed it with Nell further after that. It seems like in the majority of cases transitions between Frames of Reference would be dictated by content rather than floating point limitations?

Consider that with our system of 1 unit = 1 meter a 32 bit float can maintain accuracy of 0.5mm or better at up to 8192 meters (~5 miles). I find it hard to imagine any content developers out there who would be developing their content to stream in 5^2 mile blocks. You will almost certainly see developers breaking up their content into much smaller pieces and initializing new Frames of Reference to align with those streamed chunks rather than transitioning between systems based on precision alone.

That said it probably wouldn't be a bad idea to offer some guidance around what we feel is a recommended distance from the origin before the developer should think about moving to a new Frame of Reference. 2048 meters (1.2 miles) would be a pretty safe, conservative value IMO, given that precision at that point is still 0.1mm? Would appreciate some comments from platform owners on what they consider the minimum acceptable threshold for tracking accuracy.

(Floating point precision numbers taken from the [table on this page.](https://blog.demofox.org/2017/11/21/floating-point-precision/))

Reply:
Brandon - you bring up a good point about 5 miles being a lot, but just to be complete in our thinking I am going to ignore that for a second and assume that we want to support someone hiking across the country wearing AR glasses.

Maybe I am missing something here but I don't see how the runtime can prevent its own coordinate system from leaking to the app. If the app doesn't request a new frame of reference and the underlying system has to remap to avoid a floating point precision issue, then it is going to run into that same precision issue when it tries to convert its newly-remapped internal vectors into the app's frame-of-reference vectors, which have not been remapped. Those app vectors will be too large when being converted, right? I may be missing something...

Reply:
Maybe this ends up being related to the representation of tracking loss?

Reply:
Indeed!  Even assuming a "perfect tracking system" that has 0 tracking error and rebases its internal coordinate system every frame to a nearby origin, the user will _still_ at some point walk far enough that a distant app-created `"unbounded"` frame origin will be a cause of floating-point error when device poses are expressed in that origin.

It is therefore up to the app to absorb some reset of its `"unbounded"` frame origin to mitigate floating-point error if the device moves some number of kilometers from the origin.  The key question is whether the runtime signals a forced reset, or if the app chooses when to reset.

Since this floating-point error manifests regardless of tracking system, WebXR middleware can be designed generically to handle it at the app layer:
* An app could use a naïve strategy, just resetting its `"unbounded"` frame when the user walks some fixed distance from the origin (e.g. 1024 meters).
* Alternatively, the app could have more involved reset behavior at app-meaningful chunk boundaries, as proposed by @toji above.  An app choosing this approach benefits from being in control of the origin reset, rather than having it also forced by the runtime at arbitrary moments that differ across UA or underlying XR platform.

Presuming the app is doing one of the above, it is then up to the UA or underlying XR platform to decide how often to rebase its own internal coordinate systems, so that such well-behaved apps will not observe floating-point error.

Reply:
I would tend to agree with @thetuvix ... in unbounded frames, the coordinate system is not fixed, so the apps should be requesting to get values relative to some `XRFrameOfReference` each frame.  When the platform feels the system has moved "far enough", it can recenter.  

The whole point of requiring that all content be anchored to some FoR or CoordinateSystem is that the "current euclidean coordinate system" is only valid for a single frame;  aside from that, the coordinate system needs to be adjusted to whatever the platform says.

As has been discussed elsewhere (perhaps in the anchor repo?) this will have implications for things like managing physics.  But those can probably be abstracted once in an "XRPhysics" abstraction layer on a physics library.

Reply:
And, following @thetuvix's comment above mine (which came in as I was typing mine!), I will say that in Argon we did the reset for the user, and sent an event when the origin was reset.  We could do similarly here; have an event on an `unbounded` FoR that users could subscribe to if they wanted to know things had been rebased.  We used this to do just what he suggests: adjust all the anchored content.  The details inside the physics engine can also be adjusted there.

This will be important when geospatial content is used, as the geospatial frame should remain reasonably close to the user, since it's typically a tangent plane to the earth centered at some location.  As the user moves away from it, it gets less accurate an approximation to "the user".

Reply:
So, I'm not quite following along... are we saying that the developers need to monitor and handle when the floating point error gets too high?  Or that WebXR will do it and issue an `onreset`?

FWIW, I just moved `onreset` to the `XRFrameOfReferenceType` for another reason.

Reply:
I personally think "developers need to monitor and handle when the floating point error gets too high" is a recipe for disaster.  I think developers will want to monitor and shift the coordinate frame for more application-specific reasons (e.g., based on the distribution and layout of their content, the constraints of the frameworks they are using, etc).  

But, having the platform shift the origin of the coordinate system when appropriate to the platform (not just because of javascript numerical precision) seems appropriate.  Especially when the platform is already adjusting the coordinate systems as the user moves around, anyway, so the application layers need to deal with that.

Having an `onreset` event is useful, but optional:  "recentering" is a really large move and may require major updates to content, so it may be useful to know explicitely that this has happened.  It's easy to detect, though:  monitor one anchor and see if it's pose relative to the camera moves a large amount between frames.

Reply:
If we want WebXR to trigger an automatic reset event to the `"unbounded"` space at some defined point (e.g. 1024 meters), we could spec that out and do that consistently across all UAs.

I want to address one key point though:

@blairmacintyre:
> Especially when the platform is already adjusting the coordinate systems as the user moves around, anyway, so the application layers need to deal with that.

An inside-out tracking system will be continuously and smoothly adjusting the `"unbounded"` origin as needed to keep coordinates stable near the user.  Every time there is a tiny shift of that coordinate system, we would not trigger a `"onreset"`.  This would be hard to define anyway - when the `"unbounded"` origin shifts relative to its anchors, did the anchors move or did the frame origin move?

Given that, there will be specific known cases where various frames of reference would trigger `"onreset"`:
* The `"stationary"` frame would trigger `"onreset"` when the user engages some platform UI to reset their "local zero" origin.  Some platforms have no such UI, and apps there would never see an `"onreset"`.
* The `"bounded"` frame would trigger `"onreset"` when the user redraws their room setup, or the platform observes the user walking closer to some second stage they've defined.  Some platforms support only a single stage and do not allow for room redefinition while an app is running, and apps there would never see an `"onreset"`.
* The `"unbounded"` frame would trigger `"onreset"` when the user walks past some spec-defined boundary where we forcibly reset the app to help avoid floating-point error, or when the origin has a large discontinuity due to tracking recovery in a new map fragment.  Some platforms may not allow users to walk far enough for floating-point error, or may only support a single map fragment, and apps there would never see an `"onreset"`.

Given the above, floating-point error caused by the app being far from the `"unbounded"` origin is unlikely to be due to platform-specific tracking behavior, since it is the job of the `"unbounded"` frame to smoothly integrate that drift over time to avoid large discontinuities.  If we do want to have UAs automatically reset the `"unbounded"` frame due to floating-point error, it therefore seems that we could help keep things consistent for apps and make that a specified feature of WebXR itself.

Reply:
I am happy with the detail you added. Go forth and resolve!